[RuntimeAsync] Merge from labs, all except JIT

VSadov opened the pull request:
This PR includes everything from the runtimelab implementation of Runtime Async feature, except: JIT changes and tests. 
Some changes in between VM and JIT are hard to separate, so included here as well. 

The main effect of the VM changes is classification of methods by `MethodReturnKind`. We recognize methods that return `Task`/`ValueTask`/`Task<T>`/`ValueTask<T>` as special. 

A task-returning method definition will now get two methoddescs in the type system:
- one has the original Task-returning signature 
- another is T-returning, but called with a calling convention that uses a continuation object to handle cases of incomplete result leading to suspending/resuming the chain of calls.
 
We call the pair "variants". For most purposes the methods are identical and parallel - i.e. if generic, both have exactly same set of type parameters, if task-returning variant overrides another one in the base class, the async variant overrides the async variant in the base, etc...  
Both variants can signal that the result is not ready yet - via incomplete Task or by returning a Continuation object.

The purpose of this duality is that async variants can be much more efficient when suspension does not happen, leading to substantial savings when a chain of awaits may be turned into a chain of calls in "async space". [Here the JIT part that optimizes `Await pattern` into async calls](https://github.com/dotnet/runtime/pull/114861#discussion_r2053123450).

From the public/external point of view the async variants are an implementation detail that "does not exist", and can change in future releases. Async variants should not show up in reflection, for example.

In terms of implementation logic one variant contains the user code/IL and another is a thunk that forwards to it. Which one contains the user code depends on presence of `MethodImpl.Async`. If there is no `MethodImpl.Async` the user code belongs to the task-returning variant, otherwise to the async variant.

Note that for the most part Runtime Async feature is pay-for-play. JIT features will only engage if the user code is compiled with a compiler that is enlightened to use [the new API for async/await](https://github.com/dotnet/runtime/issues/114310). 
However, the creation of variant pairs for task-returning methods must happen unconditionally prior to possible use. Since these are just entry points in the type system, the impact should not be high. We will, of course measure and see if some action is needed on that.

In this PR the generation of variants will require `FEATURE_RUNTIME_ASYNC`, which will not be defined by default, so by default none of the above will happen.

VSadov committed:
Merge from RTL, all except JIT

VSadov commented:
CC @jakobbotsch

VSadov committed:
fix mono build

VSadov committed:
Removed public API changes.

VSadov committed:
require FEATURE_RUNTIME_ASYNC

VSadov committed:
whitespace changes

VSadov committed:
redundant change

VSadov commented:
I recommend to review this with dot `.`   Most interesting changes are under `VM`

VSadov commented on src/coreclr/vm/method.hpp:
```diff
@@ -50,11 +50,79 @@ GVAL_DECL(TADDR, g_MiniMetaDataBuffAddress);
 #define METHOD_TOKEN_RANGE_MASK ((1 << METHOD_TOKEN_RANGE_BIT_COUNT) - 1)
 
+enum class AsyncMethodKind
+{
```
Comments in `AsyncMethodKind` may be a good start with the async method classification

VSadov commented on src/coreclr/vm/clsload.cpp:
```diff
@@ -2801,7 +2801,10 @@ TypeHandle ClassLoader::PublishType(const TypeKey *pTypeKey, TypeHandle typeHnd)
+            // we associate the methoddef with the Task-returning variant since it
+            // matches the methadata signature.
+            if (!pMD->IsUnboxingStub() && !pMD->IsAsyncVariantMethod())
             {
```
This could be interesting. Both variants have the same memberDef, so it is 2->1 relationship. 
However when going from memberDef to methoddesc is 1->1 map, and starting with memberDef will produce the non-async methoddesc - the one that matches the metadata signature.  (but it may not be the one that owns the IL!)

VSadov commented on src/coreclr/vm/comutilnative.cpp:
```diff
@@ -1851,6 +1851,14 @@ FCIMPL1(MethodTable*, MethodTableNative::InstantiationArg0, MethodTable* mt);
 }
 FCIMPLEND
 
+FCIMPL1(OBJECTHANDLE, MethodTableNative::GetLoaderAllocatorHandle, MethodTable *mt)
```
When suspending, the continuation holds the state of the call. The state may need to include loader allocator reference as the method frame could be the only thing holding it alive. This helper is used in related logic.

VSadov commented on src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:
```diff
@@ -2448,6 +2448,7 @@ public TaskAwaiter GetAwaiter()
         /// </param>
         /// <returns>An object used to await this task.</returns>
+        [Intrinsic]
         public ConfiguredTaskAwaitable ConfigureAwait(bool continueOnCapturedContext)
```
ConfigureAwait can participate in JIT optimizations (not in this PR), thus needs to be matched by name.

VSadov commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -1032,6 +1042,629 @@ PCODE MethodDesc::JitCompileCodeLocked(PrepareCodeConfig* pConfig, COR_ILMETHOD_
+}
+
+void MethodDesc::EmitAsyncMethodThunk(MethodDesc* pAsyncOtherVariant, MetaSig& msig, ILStubLinker* pSL)
+{
```
Here we create a method body for an Async thunk to a Task-returning method. It is actually a fairly simple piece of code.

VSadov commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -1032,6 +1042,629 @@ PCODE MethodDesc::JitCompileCodeLocked(PrepareCodeConfig* pConfig, COR_ILMETHOD_
+        return false;
+    }
+
+    MethodDesc *pAsyncOtherVariant = this->GetAsyncOtherVariant();
```
Here we create a Task-returning variant for an Async method. It is more code than the other way around, but essentially it calls the Async method and then examines if we have a continuation returned or an actual result and calls a helper that wraps that into an appropriate `Task`.

jkotas commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -1032,6 +1042,629 @@ PCODE MethodDesc::JitCompileCodeLocked(PrepareCodeConfig* pConfig, COR_ILMETHOD_
+    return false;
+}
+
+bool MethodDesc::TryGenerateAsyncThunk(DynamicResolver** resolver, COR_ILMETHOD_DECODER** methodILDecoder)
```
Can the generation of all async helpers be in a separate file for readability? TryGenerateUnsafeAccessor was moved to a separate file for the same reason.

VSadov replied:
Moved async helpers to a different file. (managed helpers too)

jkotas commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -1569,6 +2228,12 @@ Stub * CreateInstantiatingILStub(MethodDesc* pTargetMD, void* pHiddenArg)
     // InstantiatingStub
     pCode->EmitLDC((TADDR)pHiddenArg);
 
+    // 2.3.2 Push the async continuation
```
Is there a point in numbering the comments here? Delete the numbering on all comments instead?

jkotas commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -2645,6 +2690,132 @@ HRESULT MethodTableBuilder::FindMethodDeclarationForMethodImpl(
 #pragma warning(disable:21000) // Suppress PREFast warning about overly large function
 #endif // _PREFAST_
+
+void GetNameOfTypeDefOrRef(Module* pModule, mdToken tk, LPCSTR* pName, LPCSTR* pNamespace)
```
```suggestion
static void GetNameOfTypeDefOrRef(Module* pModule, mdToken tk, LPCSTR* pName, LPCSTR* pNamespace)
```

jkotas commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -2645,6 +2690,132 @@ HRESULT MethodTableBuilder::FindMethodDeclarationForMethodImpl(
+    }
+}
+
+bool IsTypeDefOrRefImplementedInSystemModule(Module* pModule, mdToken tk)
```
```suggestion
static bool IsTypeDefOrRefImplementedInSystemModule(Module* pModule, mdToken tk)
```

jkotas commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -2645,6 +2690,132 @@ HRESULT MethodTableBuilder::FindMethodDeclarationForMethodImpl(
+{
+    if (TypeFromToken(tk) == mdtTypeRef)
+    {
+        if(!ClassLoader::ResolveTokenToTypeDefThrowing(pModule, tk, &pModule, &tk))
```
```suggestion
        if (!ClassLoader::ResolveTokenToTypeDefThrowing(pModule, tk, &pModule, &tk))
```

jkotas commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -2645,6 +2690,132 @@ HRESULT MethodTableBuilder::FindMethodDeclarationForMethodImpl(
+    return false;
+}
+
+bool IsTypeDefOrRefAByRefStruct(Module* pModule, mdToken tk)
```
```suggestion
static bool IsTypeDefOrRefAByRefStruct(Module* pModule, mdToken tk)
```

jkotas commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -2645,6 +2690,132 @@ HRESULT MethodTableBuilder::FindMethodDeclarationForMethodImpl(
+    return false;
+}
+
+MethodReturnKind ClassifyMethodReturnKind(SigPointer sig, Module* pModule, ULONG* offsetOfAsyncDetails, bool *isValueTask)
```
```suggestion
static MethodReturnKind ClassifyMethodReturnKind(SigPointer sig, Module* pModule, ULONG* offsetOfAsyncDetails, bool *isValueTask)
```

jkotas commented on src/coreclr/vm/method.hpp:
```diff
@@ -602,7 +689,8 @@ class MethodDesc
         return mcFCall == GetClassification()
-            || mcArray == GetClassification();
+            || mcArray == GetClassification()
+            || IsAsyncThunkMethod();
```
This is confusing method. It returns true for some runtime supplied methods and not others (e.g. it returns false for unboxing stubs). Is it a good idea to keep overloading it even more?

VSadov replied:
`IsRuntimeSupplied()` felt like a meaningful thing at the time. I will use the same pattern as unboxing stubs then. 
This is not the only place where we see similarities with unboxing stubs.

VSadov replied:
I've reverted `IsRuntimeSupplied()` to where it was.

jkotas commented on src/coreclr/vm/method.hpp:
```diff
@@ -803,7 +891,7 @@ class MethodDesc
         }
         CONTRACTL_END;
-        return IsIL() && !IsUnboxingStub() && GetRVA();
+        return IsIL() && !IsUnboxingStub() && GetRVA() && !IsRuntimeSupplied();
```
This is unnecessary. `GetRVA()` returns 0 for runtime supplied methods.

jkotas commented on src/coreclr/vm/method.cpp:
```diff
@@ -1171,6 +1204,9 @@ COR_ILMETHOD* MethodDesc::GetILHeader()
     }
     CONTRACTL_END
 
+    if (IsRuntimeSupplied())
```
Should this rather a be precondition to match `IsUnboxingStub`?

VSadov replied:
I've made it a precondition.  One result of that is having to add a `HasILHeader` check in couple places that eagerly call `GetILHeader()`, but would be ok with getting `NULL`.

VSadov replied:
I had to undo the precondition part. 

`HasILHeader() == false` does not prevent being asked for IL, because there are many ways how a method can provide IL, even if it has no RVA or header (i.e. IL can be dynamically generated or overriden via EnC or profiler patching), so we end up here for Async thunks. 

`IsUnboxingStub` has very special handling, so does not come here, but Async methods are treated as regular methods here, except the thunks do not provide static IL, which is then provided dynamically. 

The other changes that you mentioned are indeed unnecessary. 
As long as Async thunks report 0 in `GetRVA()`, that alone is sufficient.

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -14524,6 +14602,361 @@ bool CEEInfo::getTailCallHelpers(CORINFO_RESOLVED_TOKEN* callToken,
+    PVOID newBlob = pamTracker.Track(alloc->GetHighFrequencyHeap()->AllocMem(S_SIZE_T(sigLen)));
+    memcpy(newBlob, builderSig, sigLen);
+
+    pamTracker.SuppressRelease();
```
(multiple places has this problem, not commenting on all of them)

VSadov replied:
This is a similar pattern as used in `TailCallHelp::AllocateBlob`/`TailCallHelp::AllocateSignature`. 

I think assume the failures cannot be reasonably handled, so tracker is released right after allocation. Although I am not sure why use tracker at all in such case.

VSadov replied:
We can treat `CreateAndLinkNewILStubMethodDesc` as publish point

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:
```diff
@@ -544,6 +578,304 @@ private static unsafe void DispatchTailCalls(
             }
         }
 
+        private static Continuation AllocContinuation(Continuation prevContinuation, nuint numGCRefs, nuint dataSize)
```
All this should be in separate file (AsyncHelpers or whatever the API review decides.)

VSadov replied:
I think a good chance it will be `AsyncHelpers`

VSadov replied:
Moved to `AsyncHelpers`, which did end up the container for public API after the review.

VSadov commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -1483,6 +1484,37 @@ HCIMPL0(void, IL_Rethrow)
+
+    FC_GC_POLL_NOT_NEEDED();    // throws always open up for GC
+
+    HELPER_METHOD_FRAME_BEGIN_ATTRIB_NOPOLL(Frame::FRAME_ATTR_EXCEPTION);    // Set up a frame
```
This is very similar to `IL_Rethrow` above, just the throwable has not be thrown yet. `IL_Rethrow` sets a frame. `IL_Throw` above that also sets `Frame::FRAME_ATTR_EXCEPTION`.

I can't tell right away if `Frame::FRAME_ATTR_EXCEPTION` is easily removable from throw helpers. 
It looks like stack walker uses the presence of the throw frame in its logic.

VSadov replied:
Although `DispatchManagedException` path would not need the frame. I see what you meant.

VSadov replied:
Simply removing the frame resulted in a crash in `async\simple-eh\simple-eh.dll`. 
Looks like there is more to it than just removing the frame.

VSadov replied:
I have introduced a frameless path for everything except X86. 
We will have to fix X86 later, since it would need JIT changes as well.

VSadov replied:
The " fix X86 later" is already done. Thanks @filipnavara

jakobbotsch commented on src/coreclr/vm/codeversion.h:
```diff
@@ -61,6 +61,7 @@ class NativeCodeVersion
     NativeCodeVersionId GetVersionId() const;
     BOOL IsDefaultVersion() const;
     PCODE GetNativeCode() const;
+    PTR_PCODE GetNativeCodeSlot() const;
```
This isn't used as-is, so can be removed. It's there because I initially tried to reuse the code slot from the code version in the async resumption stub, but I was not able to make it work reliably. Instead `getAsyncResumptionStub` just allocates a new slot and later (after jitting) writes the pointer into it.

VSadov committed:
removed GetNativeCodeSlot

VSadov committed:
Apply suggestions from code review

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

VSadov committed:
fix the build

VSadov committed:
add a frameless path in IL_ThrowExact (except for X86)

VSadov committed:
keep amTracker untill after CreateAndLinkNewILStubMethodDesc

VSadov committed:
no comment numbering

jkotas commented on src/coreclr/vm/frames.cpp:
```diff
@@ -1357,7 +1357,8 @@ void HijackFrame::GcScanRoots_Impl(promote_func *fn, ScanContext* sc)
 
-    ReturnKind returnKind = m_Thread->GetHijackReturnKind();
+    bool hasAsyncRet;
+    ReturnKind returnKind = m_Thread->GetHijackReturnKind(&hasAsyncRet);
```
The async ABI should be documented in https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/clr-abi.md

VSadov replied:
Will do. It is not very complicated though. 

- There is an extra optional continuation return. If it is not NULL, means the result is not ready and the continuation has captured the state. 
The format of continuation is opaque, but changing the format will likely be R2R breaking, ... when we support R2R.

- There is also an extra continuation parameter (position depends on platform, x86 is different, as usual). 
If the continuation argument is not NULL, means this is not the first time the method runs, and it needs to unpack the continuation and continue running.
The method may return a continuation again, if there is another await that is not ready with results. That is why we do not resume the entire stack - there could be an await in a loop.

This is all there is from the ABI point, I think.

VSadov replied:
added a section on Async calling convention in the ABI document

VSadov committed:
undo changes to IsRuntimeSupplied

VSadov committed:
move async thunk generation code into separate file

VSadov committed:
Moved infrastructure helpers into CompilerServices.AsyncHelpers

VSadov committed:
Added a section on Async calling convention in ABI doc

jkotas commented on docs/design/coreclr/botr/clr-abi.md:
```diff
@@ -96,6 +96,34 @@ There is no defined/enforced/declared ordering between the generic parameter and
+
+The `Continuation` is a managed object and needs to be tracked accordingly. The GC info includes the continuation result as live at Async call sites.
+
+### Returning `Continuaton`
```
```suggestion
### Returning `Continuation`
```

jkotas commented on docs/design/coreclr/botr/clr-abi.md:
```diff
@@ -96,6 +96,34 @@ There is no defined/enforced/declared ordering between the generic parameter and
+| arm64  | x2  |
+| risc-v  | a2  |
+
+### Passing `Continuaton` argument
```
```suggestion
### Passing `Continuation` argument
```

jkotas commented on docs/design/coreclr/botr/clr-abi.md:
```diff
@@ -96,6 +96,34 @@ There is no defined/enforced/declared ordering between the generic parameter and
+
+Async calling convention is additive to other calling conventions when supported. The set of scenarios is constrained to regular static/virtual calls and does not, for example, support PInvokes or varargs. At the minimum ordinary static calls, calls with `this` parameter or generic hidden parameters are supported.
+
+Async calling convention adds an extra `Continuation` parameter and an extra return, which sematically takes precedence when not `null`. A non-null `Continuation` upon return signals that the computation is not complete and the formal result is not ready. A non-null argument means that the function is resuming and should extract the state from the `Continuation` and continue execution (while igno ...
```
Are the other arguments expected to have null/default values when Continuation argument is non-null?

VSadov replied:
The arguments need to have some value, so we do set other arguments to default values when calling with a continuation. 
I am not sure it is strictly required as they are not supposed to be used if continuation is passed in. Perhaps nothing will break if arguments contain random values. Managed refs might need to be valid though - actual objects or null, for GC reasons.

Maybe it is worth to document that arguments/returns in the presence of continuation are supposed to have default values, although I am not sure we would check/enforce that.

VSadov replied:
Actually it looks like we can return random values if non-gc return when returning continuation.

```c
void CodeGen::genReturnSuspend(GenTreeUnOp* treeNode)
{
    GenTree* op = treeNode->gtGetOp1();
    assert(op->TypeIs(TYP_REF));

    regNumber reg = genConsumeReg(op);
    inst_Mov(TYP_REF, REG_ASYNC_CONTINUATION_RET, reg, /* canSkip */ true);

    ReturnTypeDesc retTypeDesc = compiler->compRetTypeDesc;
    unsigned       numRetRegs  = retTypeDesc.GetReturnRegCount();
    for (unsigned i = 0; i < numRetRegs; i++)
    {
        if (varTypeIsGC(retTypeDesc.GetReturnRegType(i)))     <=====  zeroing only if GC
        {
            regNumber returnReg = retTypeDesc.GetABIReturnReg(i, compiler->info.compCallConv);
            instGen_Set_Reg_To_Zero(EA_PTRSIZE, returnReg);
        }
    }

    genMarkReturnGCInfo();
}
```

VSadov replied:
The resumption stub does fill arguments with default values, but that might be because it is in IL

VSadov replied:
> Maybe it is worth to document that arguments/returns in the presence of continuation are supposed to have default values, although I am not sure we would check/enforce that.

Or perhaps just mentioning that they are unused is good enough. Caller might zero them out or might not - whatever is more convenient.

jkotas replied:
Have we considered generating the resumption stub in the JIT? It feels like we are leaving some perf and code size on the table here.

VSadov replied:
It is possible, although resumption is not a hot path. A lot happens between suspension and resumption.

I think every part will be revisited eventually with perf in mind, but this may not be the most impactful area. 
This is where, I think, getting correct and reasonably fast behavior is good enough for now.

jakobbotsch replied:
I think eventually we will want to add support for generating multiple entry points for async functions and have the JIT generate an entry point for resumption and an entry point for fresh calls. It will also remove the extra null argument being passed on the synchronous path.

This is the 4th item under https://github.com/dotnet/runtimelab/blob/feature/async2-experiment/docs/design/features/runtime-handled-tasks.md#potential-future-improvements

jakobbotsch replied:
It's not totally trivial to completely remove the resumption stub, even with multiple entry points -- currently the resumption stub has the important job of setting up space for the arguments. The resumed function uses that space to store parameters from the continuation when parameters are live. Particularly for implicit byrefs that space exists on the stack frame of the resumption stub and cannot easily be allocated by the callee (without deoptimizing e.g. tailcalls).

VSadov commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -14477,6 +14555,360 @@ bool CEEInfo::getTailCallHelpers(CORINFO_RESOLVED_TOKEN* callToken,
+    ILCodeStream* pCode = sl.NewCodeStream(ILStubLinker::kDispatch);
+
+    int numArgs = 0;
+
```
The resumption stub is called with a single continuation argument. 
```cs
public delegate*<Continuation, Continuation?> Resume;
```

To call the actual Async method we set all arguments to `default`/`null`, starting with `this`, if present.

VSadov commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -14477,6 +14555,360 @@ bool CEEInfo::getTailCallHelpers(CORINFO_RESOLVED_TOKEN* callToken,
+    }
+
+    // Continuation
+    pCode->EmitLDARG(0);
```
The only argument that has value is the continuation argument, we set it here.

VSadov commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -14477,6 +14555,360 @@ bool CEEInfo::getTailCallHelpers(CORINFO_RESOLVED_TOKEN* callToken,
+        pCode->EmitLDIND_I();
+    }
+
+    pCode->EmitCALLI(pCode->GetSigToken(calliSig.GetRawSig(), calliSig.GetRawSigLen()), numArgs, msig.IsReturnTypeVoid() ? 0 : 1);
```
And then we `CALLI` the actual Async method. 
I think nothing would break if other arguments had random values, if non-GC. But it so happens that we clear them.

VSadov replied:
That just comes from the fact that if some parameter was used/live after resuming, it's value would have been captured in the continuation and then restored, which would override whatever we'd pass in.

VSadov committed:
Apply suggestions from code review

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

jakobbotsch commented on src/coreclr/vm/stubgen.cpp:
```diff
@@ -819,10 +834,10 @@ size_t ILStubLinker::Link(UINT* puMaxStack)
 
 #ifdef _DEBUG
-    if (fStackUnderflow)
+    if (fStackUnderflow || this->GetNumEHClauses() != 0)
     {
         LogILStub(CORJIT_FLAGS());
-        CONSISTENCY_CHECK_MSG(false, "IL stack underflow! -- see logging output");
+//        CONSISTENCY_CHECK_MSG(false, "IL stack underflow! -- see logging output");
     }
 #endif // _DEBUG
```
We should undo this change, I think it is just an old debug change accidentally committed

VSadov replied:
This has been undone.

VSadov committed:
undo unnecessary change in stubgen

VSadov commented:
rebased onto most recent main to resolve conflicts

VSadov committed:
Use the newest API

VSadov committed:
Use approved value for MethodImplAttributes.Async

VSadov commented:
Now this PR includes the public API as well. As approved in https://github.com/dotnet/runtime/issues/114310

VSadov commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -7797,12 +7799,36 @@ MethodDesc* MethodTable::GetParallelMethodDesc(MethodDesc* pDefMD)
+    else
+    {
+        // Slow path for finding the Async variant (or not-Async variant, if we start from Async one)
+        // This could be optimized with some trickery around slot numbers, but doing so is ... confusing, so I'm not implementing this yet
```
A solution for this was prototyped in https://github.com/dotnet/runtimelab/pull/3086 

It works, but it is a space-for-speed tradeoff of a kind that we typically do only for something that impacts run time (vs. JIT time), so we'd need more perf data to see if such change is a necessary or what we have is good enough.

VSadov replied:
A less intrusive change for cases where both variants live in the same chunk is also possible. 

Either way it is better to do perf tuning when we can measure the impact on something closer to the realworld app. We may have other areas more likely to move the needle.

filipnavara committed:
Implement IL_ThrowExact for x86/funclets

jkotas commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -1032,6 +1042,21 @@ PCODE MethodDesc::JitCompileCodeLocked(PrepareCodeConfig* pConfig, COR_ILMETHOD_
 }
 
+bool MethodDesc::TryGenerateTransientILImplementation(DynamicResolver** resolver, COR_ILMETHOD_DECODER** methodILDecoder)
+{
```
```suggestion
{
    STANDARD_VM_CONTRACT;

```

jkotas commented on src/coreclr/vm/clrtocomcall.cpp:
```diff
@@ -230,6 +230,11 @@ I4ARRAYREF SetUpWrapperInfo(MethodDesc *pMD)
 
         GCX_PREEMP();
 
+        if (pMD->IsAsyncMethod())
```
Yes, many of these can be hit. For example, this throws NotSupportedException in the runtimelab async branch and prints 7 in main:

```
using System;
using System.Runtime.InteropServices;
using System.Threading.Tasks;

Console.WriteLine(Marshal.GetEndComSlot(typeof(IFoo)));

public interface IFoo
{
    Task M();
}
```

Before we turn on runtime async support in main, we need to go over all places where you are adding these NotSupportedException, validate whether they can be hit, and fix things up as appropriate to avoid breaking changes. We know that the test coverage for built-in COM interop is very spotty, but customers depend on every corner case behavior of built-in COM interop. cc @AaronRobinsonMSFT 

Could you please tag all these places that need review with async TODO so that they can be found easily and none of them falls through the cracks?

AaronRobinsonMSFT commented on src/coreclr/inc/corhdr.h:
```diff
@@ -640,13 +640,15 @@ typedef enum CorMethodImpl
     miNoOptimization     =   0x0040,   // Method may not be optimized.
     miAggressiveOptimization = 0x0200, // Method may contain hot code and should be aggressively optimized.
 
+    miAsync              =   0x2000,   // Method requires async state machine rewrite.
```
We also need to update the ECMA-335 augments document - `docs/design/specs/Ecma-335-Augments.md`.

VSadov replied:
The changes to ECMA-335 are tracked separately in https://github.com/dotnet/runtime/blob/main/docs/design/specs/runtime-async.md

VSadov replied:
> We will need to add support for this to ilasm/ildasm. Is it tracked somewhere?

https://github.com/dotnet/runtime/issues/115093

filipnavara replied:
The flag value in the .md above needs to be updated (0x400 vs 0x2000)

filipnavara committed:
Addres PR feedback

davidwrighton reviewed (changes requested):
In `RuntimeMethodHandle::GetMethodDef` there is a call to GetCallingConventionInfo. we should check what impact that has.

Otherwise it looks pretty good to me. The only thing I think really needs to be fixed is the issue with integer overflow on the cast to SLOT_INDEX on cMethAndGaps and multiplying it by 2. I think that needs to be under the ifdef until we figure out how to make that safe.

VSadov commented:
> Could you please tag all these places that need review with async TODO so that they can be found easily and none of them falls through the cracks?

Will do. Makes sense. I'll log a tracking issue and mark all the places where we check for Async because we do not have a well defined behavior for that case.

VSadov commented on src/coreclr/vm/class.cpp:
```diff
@@ -679,6 +679,7 @@ HRESULT EEClass::AddMethodDesc(
                                                             classification,
                                                             TRUE, // fNonVtableSlot
                                                             TRUE, // fNativeCodeSlot
+                                                            FALSE, /* HasAsyncMethodData */
```
Supporting EnC is likely to be asked for. Right now we do not, so our only choice is to fail somehow when adding a variant method.

Do we have a precedent when we do not allow EnC certain methods? Like do we allow EnC for `Synchronized` ?

What happens when EnC modifies something that needs unboxing sub?

jkotas replied:
> Do we have a precedent when we do not allow EnC certain methods?

Roslyn is the primary place where the EnC limitations are enforced. The enforcement is done based on EnC capabilities of current runtime fetched by https://github.com/dotnet/runtime/blob/32463ba7daff98330c8bc550280c6dc26538de08/src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdater.cs#L58.

The runtime enforcement of EnC limitations is opportunistic. We do not try to catch the unsupported EnC scenarios holistically in the runtime.

jkotas replied:
> do we allow EnC for Synchronized ?

I think so. FixContextForEnC that remaps the old method frame to the new one talks about synchronized methods: https://github.com/dotnet/runtime/blob/32463ba7daff98330c8bc550280c6dc26538de08/src/coreclr/vm/eetwain.cpp#L285 . 

> What happens when EnC modifies something that needs unboxing sub?

Unboxing stubs do not need vtable slot. Unboxing stubs can live in a side-table and be created on demand by the code in getmeth.cpp (look for TightlyBound vs. the other path).

jkotas replied:
> Supporting EnC is likely to be asked for.

Yes, EnC will need to be dealt with before we ship this for real. Add it to the punch list if it is not there already.

jakobbotsch replied:
Another case that is not supported for EnC and which Roslyn knows about intrinsically is methods with stackalloc in them. I suppose until we have this support Roslyn could similarly disallow edits in runtime async methods.

VSadov replied:
I've logged an issue on EnC - https://github.com/dotnet/runtime/issues/115095

We may need to think about other scenarios when methods are dynamically overriden/patched/versioned too (like profiler).

VSadov commented on src/coreclr/vm/commodule.cpp:
```diff
@@ -302,7 +302,13 @@ extern "C" INT32 QCALLTYPE ModuleBuilder_GetMemberRefOfMethodInfo(QCall::ModuleH
     }
 
-    if (pMeth->GetMethodTable()->GetModule() == pModule)
+    if (pMeth->IsAsyncVariantMethod())
```
Yes, when plainly asking for a methoddesc representing a token it should always go to the Task-returning variant as that is the thing we have in metadata. For other variant need to ask explicitly.

Making a delegate of a task-returning method and calling should work as a test, right.

VSadov replied:
I've added a direct scenario to test this in reflection\ldtoken\ldtokenmember. The scenario is passing with or without `FEATURE_RUNTIME_ASYNC`

Returning wrong method for ldtoken would be very breaking in general. It'll likely would break xunit itself.

VSadov commented on src/coreclr/vm/readytoruninfo.cpp:
```diff
@@ -1077,6 +1079,10 @@ bool ReadyToRunInfo::GetPgoInstrumentationData(MethodDesc * pMD, BYTE** pAllocat
     if (ReadyToRunCodeDisabled())
         return false;
 
+    // TODO: (async) PGO support for async variants
```
Yes, we should replace TODOs with issues. 
Except for the "this most likely unreacahble with Async, otherwise do not know what to".  The latter are ok to be TODOs and one umbrella issue to review them.

am11 replied:
Perhaps `TODO-Async:` is a better, and an easier-to-grep, tag.

VSadov replied:
`// TODO: (async) ` is used already in many places, I'll continue with that

VSadov replied:
Tracking issue for PGO support: https://github.com/dotnet/runtime/issues/115096

VSadov commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -2700,7 +2871,7 @@ MethodTableBuilder::EnumerateClassMethods()
         BuildMethodTableThrowException(IDS_CLASSLOAD_TOO_MANY_METHODS);
 
-    bmtMethod->m_cMaxDeclaredMethods = (SLOT_INDEX)cMethAndGaps;
+    bmtMethod->m_cMaxDeclaredMethods = (SLOT_INDEX)cMethAndGaps * 2;
```
Need to think about this. Good catch. 
I can sweep it under `FEATURE_RUNTIME_ASYNC ` for now, but eventually we will need a way to prevent overflows.

VSadov commented:
> In RuntimeMethodHandle::GetMethodDef there is a call to GetCallingConventionInfo. we should check what impact that has.

I see `GetCallingConventionInfo` under `Signature_Init`, is that the one?

I'd assume we never see Async variants as an input, but can see async helpers (as in `Await`). I think we can produce a truthful signature for those, but no need to make Async call convention public. Ultimately Async methods are not callable via reflection and should throw something. 
I think there might be existing scenarios where user attempts to call something that is a "real method, but can't be called directly". Non-variant Async methods can be modeled in a similar way.

VSadov committed:
added some TODOs

VSadov committed:
Update src/coreclr/vm/prestub.cpp

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

VSadov committed:
some simple PR feedback

am11 commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -1484,6 +1485,47 @@ HCIMPL0(void, IL_Rethrow)
+    SoftwareExceptionFrame exceptionFrame;
+#ifdef TARGET_X86
+    _ASSERTE(!"Introduce `IL_ThrowExact_x86` following the pattern of other Throw helpers.");
+    // exceptionFrame.UpdateContextFromTransitionBlock(transitionBlock);
```
Merging main and cherry-picking https://github.com/dotnet/runtimelab/commit/ad964673f457fde7bed7634fad7a7f9644a2edf8 would resolve it.

VSadov replied:
This has been resoved now.

VSadov committed:
Fail if EnC targets an Async method

jkotas commented on src/coreclr/vm/class.cpp:
```diff
@@ -661,6 +661,17 @@ HRESULT EEClass::AddMethodDesc(
+    ULONG offsetOfAsyncDetails;
+    bool isValueTask;
+    MethodReturnKind returnKind = ClassifyMethodReturnKind(sigParser, pModule, &offsetOfAsyncDetails, &isValueTask);
+    if (returnKind != MethodReturnKind::NormalMethod)
```
Should this have TODO Async? 

This will need to be fixed before `FEATURE_RUNTIME_ASYNC` is enabled in dotnet/runtime.

VSadov replied:
I am going to log an issue on EnC implementation with Async, but can have TODOs as well.

VSadov replied:
I've added TODOs and here is a tracking issue on EnC: https://github.com/dotnet/runtime/issues/115095

jkotas commented on src/coreclr/vm/encee.cpp:
```diff
@@ -344,6 +344,12 @@ HRESULT EditAndContinueModule::UpdateMethod(MethodDesc *pMethod)
 
+    if (pMethod->HasAsyncMethodData())
+    {
+        LOG((LF_ENC, LL_INFO100, "**Error** EnC for Async methods is NYI"));
```
Same here

VSadov committed:
moved IsEnCAddedMethod bit to Flags4

VSadov committed:
couple TODOs for EnC NYI

VSadov commented on src/coreclr/vm/method.hpp:
```diff
@@ -1663,8 +1760,11 @@ class MethodDesc
+        // Method was added via Edit And Continue
+        enum_flag3_EnCAddedMethod                           = 0x0800,
+
         // enum_flag3_HasPrecode implies that enum_flag3_HasStableEntryPoint is set.
```
moved to enum_flag4

VSadov committed:
add a simple ldtoken test scenario for Task-returning method

VSadov commented on src/coreclr/vm/asyncthunks.cpp:
```diff
@@ -0,0 +1,620 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// ===========================================================================
+// File: Prestub.cpp
```
updated

VSadov commented:
> Could you please tag all these places that need review with async TODO so that they can be found easily and none of them falls through the cracks?

Added `// TODO: (async) revisit and examine if this needs to be supported` and a tracking issue to revisit/remove the TODOs: https://github.com/dotnet/runtime/issues/115094

VSadov commented on src/coreclr/vm/multicorejit.cpp:
```diff
@@ -398,6 +398,12 @@ HRESULT MulticoreJitRecorder::WriteOutput(IStream * pStream)
         MethodDesc * pMethod = m_JitInfoArray[i].GetMethodDescAndClean();
+        if (pMethod->IsAsyncVariantMethod())
+        {
+            // TODO: (async) consider adding support for async variants in the future
```
Tracking issue on Multicore JIT: https://github.com/dotnet/runtime/issues/115097

VSadov committed:
Put back BypassReadyToRun untill we are sure about R2R support

VSadov commented on src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncHelpers.cs:
```diff
@@ -0,0 +1,172 @@
+        // "BypassReadyToRun" is until AOT/R2R typesystem has support for MethodImpl.Async
+        // Must be NoInlining because we use AsyncSuspend to manufacture an explicit suspension point.
+        // It will not capture/restore any local state that is live across it.
+        [BypassReadyToRun]
```
Removing `[BypassReadyToRun]` breaks some scenarios. I have put it back until R2R typesystem at least understands `MethodImpl.Async` and its impact on call convention.

The followup issue for R2R is here: https://github.com/dotnet/runtime/issues/115098

VSadov commented:
I think I have addressed all suggestions/concerns, but there were many, so I may have missed some. 
@jkotas @davidwrighton - please take a look again if there is something more needed. Thanks!

davidwrighton reviewed (approved):
Looks good to me, but please get an approval from Jan too.

jkotas reviewed (approved):
LGTM as well!

VSadov commented:
Thanks!!!

max-charlamb commented on src/coreclr/vm/arm64/stubs.cpp:
```diff
@@ -768,9 +768,11 @@ void HijackFrame::UpdateRegDisplay_Impl(const PREGDISPLAY pRD, bool updateFloats
 
      pRD->volatileCurrContextPointers.X0 = &m_Args->X0;
      pRD->volatileCurrContextPointers.X1 = &m_Args->X1;
+     pRD->volatileCurrContextPointers.X1 = &m_Args->X2;
```
Is this intended to overwrite the value written above? It looks like this should be:
```suggestion
     pRD->volatileCurrContextPointers.X2 = &m_Args->X2;
```

VSadov replied:
Right. Looks like a typo. Thanks!
